app/Http/Controllers/BeatmapDiscussionsController.php

```diff
@@ -120,6 +123,129 @@ public function restore($id)
+        }
+
+        $output = [];
+        try {
+            DB::beginTransaction();
+
+            // create the issues for the embeds first
+            $childIds = [];
+            $blockCount = 0;
+            foreach ($document as $block) {
+                if (!isset($block['type'])) {
+                    throw new \Exception(trans('beatmap_discussions.review.validation.invalid_block_type'));
+                }
+                switch ($block['type']) {
+                    case 'embed':
+                        $message = $block['text'];
+                        $beatmapId = $block['beatmapId'] ?? null;
+
+                        $discussion = new BeatmapDiscussion([
+                            'beatmapset_id' => $beatmapset->getKey(),
+                            'user_id' => Auth::user()->getKey(),
+                            'resolved' => false,
+                            'message_type' => $block['discussionType'],
+                            'beatmap_id' => $beatmapId,
+                        ]);
+                        $discussion->saveOrExplode();
+
+                        $postParams = [
+                            'user_id' => Auth::user()->user_id,
+                            'message' => $message,
+                        ];
+                        $post = new BeatmapDiscussionPost($postParams);
+                        $post->beatmapDiscussion()->associate($discussion);
+                        $post->saveOrExplode();
+
+                        $issues[] = [
+                            'discussion' => $discussion->getKey(),
+                            'post' => $post->getKey(),
+                        ];
+                        $childIds[] = $discussion->getKey();
+                        break;
+                }
+                $blockCount++;
+            }
+
+            $minIssues = config('osu.beatmapset.discussion_review_min_issues');
+            if (empty($childIds) || count($childIds) < $minIssues) {
+                return error_popup(trans_choice('beatmap_discussions.review.validation.minimum_issues', $minIssues), 422);
+            }
+
+            $maxBlocks = config('osu.beatmapset.discussion_review_max_blocks');
+            if ($blockCount > $maxBlocks) {
+                return error_popup(trans_choice('beatmap_discussions.review.validation.too_many_blocks', $maxBlocks), 422);
+            }
+
+            // generate the post body now that the issues have been created
+            foreach ($document as $block) {
+                switch ($block['type']) {
+                    case 'paragraph':
+                        if (!$block['text']) {
+                            throw new \Exception(trans('beatmap_discussions.review.validation.paragraph_missing_text'));
+                        }
+                        // escape embed injection attempts
+                        $text = preg_replace('/%\[\]\(#(\d+)\)/', '%\[\]\(#$1\)', $block['text']);
+                        $output[] = "{$text}\n";
```

Comment:
I wonder if storing the document as is (with `embed` changed into `embed_link` or something) would be easier.